dipy/reconst/weights_method.py

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/python
+""" Functions for defining weights for iterative fitting methods."""
+
+import numpy as np
+
+MIN_POSITIVE_SIGNAL = 0.0001
+
+
+def simple_cutoff(residuals, log_residuals, pred_sig,
+                  design_matrix, leverages, C, cutoff):
+    """ Define outliers based on signal.
+    """
+    leverages[np.isclose(leverages, 1.0)] = 0.99  # avoids rare issues
+    HAT_factor = np.sqrt(1 - leverages)
+    cond = (np.abs(residuals) > + cutoff * C * HAT_factor)
+    return cond
+
+
+def two_eyes_cutoff(residuals, log_residuals, pred_sig,
+                    design_matrix, leverages, C, cutoff):
+    """ Two-eyes approach to define outliers, see [1]_.
+
+    References
+    ----------
+    .. [1] Collier, Q., Veraart, J., Jeurissen, B., den Dekker, A., Sijbers,
+       J., 2015. Iterative reweighted linear least squares for accurate, fast,
+       and robust estimation of diffusion magnetic resonance parameters: IRLLS
+       for Estimation of Diffusion MR Parameters. Magnetic Resonance in
+       Medicine 73, 2174-2184.
+    """
+    leverages[np.isclose(leverages, 1.0)] = 0.99  # avoids rare issues
+    HAT_factor = np.sqrt(1 - leverages)
+    cond = (residuals > + cutoff * C * HAT_factor) |\
+           (log_residuals < - cutoff * C * HAT_factor / pred_sig)
+    return cond
+
+
+def weights_method_wls_m_est(data, pred_sig, design_matrix, leverages,
+                             idx, total_idx, last_robust,
+                             m_est="gm", cutoff=3,
+                             outlier_condition_func=simple_cutoff):
+    """ Calculate M-estimator weights for WLS model.
+
+    Parameters
+    ----------
+    data : array
+        The measured signal.
+    pred_sig : array
+        The predicted signal, given a previous fit.
+        Has the same shape as data.
+    design_matrix : array (g, ...)
+        Design matrix holding the covariants used to solve for the
+        regression coefficients.
+    leverages : array
+        The leverages (diagonal of the 'hat matrix') of the fit.
+    idx : int
+        The current iteration number.
+    total_idx : int
+        The total number of iterations.
+    last_robust : array (int),
+        Array of 1s (inlier) and 0s (outlier).
+        Has the same shape as data.
+    m_est : which M-estimator weighting scheme to use. Currently,
```

Comment:
`str, optional` then description

Reply:
Revise `last_robust` parameter type: if it 1s and 0s, then why not a boolean? and docstring: use ndarray, remove the type, and remove the comma. Make "Has the same shape as data" dwell on the same line as the previous line (and have not checked the doc, but if it needs to have the same shape, is this checked in the code?).

The `m_est` docstring will most likely be rendered incorrectly: use only the type (missing as it stands now) on the same line as the parameter name, add the description on a separate line.

These apply to the entire PR.

Reply:
well, it's a boolean array. Are you only saying that the docstring is wrong?

Reply:
If it's a boolean array, then the docstring should read something like:
```
last_robust : ndarray 
    True for inlier indices and False for outlier indices. Must have the same shape as data.
```

The use of ndarray applies to all array docstrings.

The use of booleans is reaffirmed by this line here using a boolean NumPy function:
https://github.com/dipy/dipy/pull/3358/files#diff-a7d1f52350ccb4a2b7ff4decac1456106cda4471627ec532985e7e52c2cfc2b2R117

Assuming that these will be made true booleans, these blocks:
https://github.com/dipy/dipy/pull/3358/files#diff-a7d1f52350ccb4a2b7ff4decac1456106cda4471627ec532985e7e52c2cfc2b2R119-R120
and
https://github.com/dipy/dipy/pull/3358/files#diff-a7d1f52350ccb4a2b7ff4decac1456106cda4471627ec532985e7e52c2cfc2b2R124-R125

can then be simplified to use true boolean indexing:
```
        w[~robust] = 0.0
        w[robust] = 1.0
```

and
```
        w[~robust] = 0.0
        w[robust] = pred_sig[robust == 1]**2
````

Please add a check on the dimensionality of the array at the beginning of the method.

Thanks for considering this and for all the endeavor and persistence Sam.

Reply:
Gotcha, this makes sense, I'll make the changes.